ocaml/src/asd_io_scheduler.ml

```diff
@@ -105,50 +119,152 @@ let run t ~fsync ~fs_fd =
+
+             let blobs_count =
+               List.fold_left
+                 (fun acc (blobs, _) ->
+                  acc + List.length blobs)
+                 0
+                 acc
+             in
+
+             let module FL = Lwt_extra2.FoldingCountDownLatch in
+             let latch = FL.create
+                           ~count:blobs_count
+                           ~acc:[]
+                           ~f:(fun acc parent_dir ->
+                               if List.mem parent_dir acc
+                               then acc
+                               else parent_dir :: acc)
+             in
+
+             let sync_parent_dirs_t =
+               if fsync
+               then
+                 begin
+                   FL.await latch >>= fun parent_dirs ->
+                   Lwt_log.debug_f
+                     "Syncing parent dirs %s"
+                     ([%show : string list] parent_dirs) >>= fun () ->
+                   Lwt_list.iter_p
+                     (fun parent_dir ->
+                      Lwt_extra2.fsync_dir parent_dir)
+                     parent_dirs
+                 end
+               else
+                 Lwt.return ()
+             in
+
+             Lwt_list.map_p
+               (fun (blobs, waker) ->
+                Lwt_list.iter_p
+                  (fun (fnr, blob) ->
+                   t.write_blob
+                     fnr blob
+                     (fun fd parent_dir ->
+
+                      if fsync
+                      then
+                        begin
+                          (* wait for all the other blobs to be pushed to
+                           * the filesystem too before syncing *)
+                          FL.notify latch parent_dir;
+                          FL.await latch >>= fun _ ->
+
+                          Lwt_unix.fsync fd
+                        end
+                      else
+                        Lwt.return_unit))
+                  blobs >>= fun () ->
+                Lwt.return waker)
+               acc >>= fun wakers ->
+
+             sync_parent_dirs_t >>= fun () ->
+
+             List.iter
+               (fun (_, waker) ->
+                Lwt.wakeup_later waker ())
+               acc;
+
+             Lwt.return_unit) >>= fun (t_fsyncs, ()) ->
+
+          let logger =
+            if t_fsyncs < 0.5 then Lwt_log.debug_f
+            else if t_fsyncs < 4.0 then Lwt_log.info_f
+            else Lwt_log.warning_f
+          in
+          logger "writing blobs + syncing(%b) took %f for %i waiters, cost %i"
+                 fsync t_fsyncs waiters_len cost
+        end
+      else
+        Lwt.return_unit
     end >>= fun () ->
 
+    (if Lwt_buffer.has_item t.rocksdb_sync
+     then
+       begin
+         Lwt_buffer.harvest t.rocksdb_sync >>= fun wakers ->
+         t.sync_rocksdb ();
+         List.iter
+           (fun waker -> Lwt.wakeup_later waker ())
+           wakers;
+         Lwt.return_unit
+       end
+     else
```

Comment:
Anytime we get here it could be empty (due to the pick on wait_for_item above). I added the check so we wouldn't have to issue the sync to rocksdb unnecessarily